README.md
# SilverStripe Instagram
[![Build Status](https://travis-ci.org/SomarDesignStudios/silverstripe-instagram.svg?branch=master)](https://travis-ci.org/SomarDesignStudios/silverstripe-instagram)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/SomarDesignStudios/silverstripe-instagram/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/SomarDesignStudios/silverstripe-instagram/?branch=master)

Add an Instagram gallery to any content page.

## Requirements

- PHP 5.5.9+
- SSL
- cURL
- SilverStripe 3.2+
- An Instagram client ID and secret

## Features

- Create Instagram galleries using a Shortcode `[instagram username="somardesignstudios"]`
- Integration with TinyMCE for easy Shortcode insertion
- Multiple account support - create galleries from multiple Instagram accounts.
- OAuth authentication from within the CMS.
- Images and videos are cached using `SS-Cache` to reduce API requests.

## Documentation

- [Developer docs](docs/en/index.md)
- [User guide](docs/en/user-guide.md)



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

## v1.0.0

- Integration with TinyMCE
- Multiple account support
- OAuth authentication from within the CMS
- Caching via default `SS_Cache::factory`
- Basic "load more" paging
- Developer Docs and User Guide


tests/InstagramAdminTest.php
<?php

class InstagramAdminTest extends SapphireTest
{
    public function testGetInstagramAccountIDFromSession()
    {

    }

    public function testOAuth()
    {

    }
}


tests/InstagramAccountTest.php
<?php

use \Mockery as Mockery;

class InstagramAccountTest extends SapphireTest
{
    public function tearDown()
    {
        Mockery::close();
        parent::tearDown();
    }

    protected function getNewInstagramAccountMock()
    {
        return Mockery::mock('InstagramAccount');
    }

    public function testUpdateAccessToken()
    {
        $validToken =
            '{"user":{"username":"somardesignstudios","bio":"","website":"","profile_picture":"",' .
            '"full_name":"","id":"12345678"},"access_token":"","resource_owner_id":""}';

        $invalidToken =
            '{"user":{"username":"somardesignstudios","bio":"","website":"","profile_picture":"",' .
            '"full_name":"","id":"87654321"},"access_token":"","resource_owner_id":""}';

        /*
         * Should throw if the session state is invalid.
         */
        $mock = $this->getNewInstagramAccountMock()->makePartial();
        $mock
            ->shouldReceive('getSessionOAuthState')
            ->once()
            ->withNoArgs()
            ->andReturn('state');

        try {
            $mock->updateAccessToken($validToken, 'invalidState');
        } catch (Exception $e) {
            $this->assertEquals('Trying to set token on wrong InstagramAccount', $e->getMessage());
        }

        /*
         * Should throw if the session state is valid and the token username doesn't match the title.
         */
        $mock = $this->getNewInstagramAccountMock()->makePartial();
        $mock
            ->shouldReceive('getSessionOAuthState')
            ->once()
            ->withNoArgs()
            ->andReturn('state');
        $mock
            ->shouldReceive('getField')
            ->once()
            ->with('Title')
            ->andReturn('otherusername');

        try {
            $mock->updateAccessToken($validToken, 'state');
        } catch (Exception $e) {
            $this->assertEquals('Trying to set token on wrong InstagramAccount', $e->getMessage());
        }

        /*
         * Should set the token if session state is valid and there's no current token.
         */
        $mock = $this->getNewInstagramAccountMock()->makePartial();
        $mock
            ->shouldReceive('getSessionOAuthState')
            ->once()
            ->withNoArgs()
            ->andReturn('state');
        $mock
            ->shouldReceive('getField')
            ->with('Title')
            ->andReturn('somardesignstudios');
        $mock
            ->shouldReceive('getField')
            ->with('AccessToken')
            ->andReturn(null);
        $mock
            ->shouldReceive('setField')
            ->once()
            ->with('AccessToken', $validToken);

        $mock->updateAccessToken($validToken, 'state');

        /*
         * Should throw if the session state is valid, the token username matches the title,
         * and the new token is for another account.
         */
        $mock = $this->getNewInstagramAccountMock()->makePartial();
        $mock
            ->shouldReceive('getSessionOAuthState')
            ->once()
            ->withNoArgs()
            ->andReturn('state');
        $mock
            ->shouldReceive('getField')
            ->with('Title')
            ->andReturn('somardesignstudios');
        $mock
            ->shouldReceive('getField')
            ->with('AccessToken')
            ->andReturn($validToken);

        try {
            $mock->updateAccessToken($invalidToken, 'state');
        } catch (Exception $e) {
            $this->assertEquals('Trying to set token on wrong InstagramAccount', $e->getMessage());
        }

        /*
         * Should update the token if the session state is valid and the token IDs match.
         */
        $mock = $this->getNewInstagramAccountMock()->makePartial();
        $mock
            ->shouldReceive('getSessionOAuthState')
            ->once()
            ->withNoArgs()
            ->andReturn('state');
        $mock
            ->shouldReceive('getField')
            ->with('Title')
            ->andReturn('somardesignstudios');
        $mock
            ->shouldReceive('getField')
            ->with('AccessToken')
            ->andReturn($validToken);
        $mock
            ->shouldReceive('setField')
            ->once()
            ->with('AccessToken', $validToken);

        $mock->updateAccessToken($validToken, 'state');
    }

    public function testGetOAuthStateValueFromLoginURL()
    {
        $instagramAccount = new InstagramAccount();

        /*
         * Should return null if no URL is passed.
         */
        $this->assertEquals(null, $instagramAccount->getOAuthStateValueFromLoginURL());

        /*
         * Should return null if there's no "state" query string.
         */
        $this->assertEquals(
            null,
            $instagramAccount->getOAuthStateValueFromLoginURL(
                'http://example.com/admin/admin/instagram/InstagramAccount/OAuth?code=123'
            )
        );

        /*
         * Should return the value of the "state" query string if it exists.
         */
        $this->assertEquals(
            'abc',
            $instagramAccount->getOAuthStateValueFromLoginURL(
                'http://example.com/admin/admin/instagram/InstagramAccount/OAuth?code=123&state=abc'
            )
        );
    }

    public function testIsValidMediaID()
    {
        /*
         * Should return false if no media ID is passed.
         */
        $instagramAccount = new InstagramAccount();

        $this->assertEquals(false, $instagramAccount->isValidMediaID());

        /*
         * Should return false if the account ID is unavailable.
         */
        $mock = $this->getNewInstagramAccountMock()->makePartial();
        $mock
            ->shouldReceive('getInstagramID')
            ->withNoArgs()
            ->andReturn(null);

        $this->assertEquals(
            false,
            $mock->isValidMediaID('123456789012345678_123')
        );

        /*
         * Should return false if the media ID is less than 18 digits.
         */
        $mock = $this->getNewInstagramAccountMock()->makePartial();
        $mock
            ->shouldReceive('getInstagramID')
            ->withNoArgs()
            ->andReturn('123');

        $this->assertEquals(
            false,
            $mock->isValidMediaID('12345678901234567_123')
        );

        /*
         * Should return false if the media ID is greater than 19 digits.
         */
        $this->assertEquals(
            false,
            $mock->isValidMediaID('12345678901234567890_123')
        );

        /*
         * Should return false if the media ID length is valid but contains non-digits.
         */
        $this->assertEquals(
            false,
            $mock->isValidMediaID('123456789Z12345678_123')
        );

        $this->assertEquals(
            false,
            $mock->isValidMediaID('123456789&12345678_123')
        );

        /*
         * Should return false if a valid media ID is not followed by an underscore.
         */
        $this->assertEquals(
            false,
            $mock->isValidMediaID('123456789012345678-123')
        );

        /*
         * Should return false if the media ID doesn't end with the account ID.
         */
        $this->assertEquals(
            false,
            $mock->isValidMediaID('123456789012345678_321')
        );

        /*
         * Should return true if the media ID is 18 or 19 characters followed by
         * an underscore and the account ID.
         */
        $this->assertEquals(
            true,
            $mock->isValidMediaID('123456789012345678_123')
        );

        $this->assertEquals(
            true,
            $mock->isValidMediaID('1234567890123456789_123')
        );
    }
}


tests/InstagramPageExtensionTest.php
<?php

class InstagramPageExtensionTest extends SapphireTest
{
    public function testInstagramShortcodeHandler()
    {

    }
}


code/admin/InstagramAdmin.php
<?php

use League\OAuth2\Client\Provider\Exception\InstagramIdentityProviderException;

class InstagramAdmin extends ModelAdmin
{
    /**
     * @var string
     */
    private static $url_segment = 'instagram';

    /**
     * @var string
     */
    private static $menu_title = 'Instagram';

    /**
     * @var array
     */
    private static $managed_models = [
        'InstagramAccount',
    ];

    /**
     * @var array
     */
    private static $allowed_actions = [
        'ImportForm',
        'SearchForm',
        'OAuth',
    ];

    /**
     * @var string
     */
    private static $menu_icon = 'silverstripe-instagram/images/instagram-logo.png';

    /**
     * @param int $id
     * @param FieldList $fields
     * @return Form
     */
    public function getEditForm($id = null, $fields = null)
    {
        $form = parent::getEditForm($id, $fields);

        $gridFieldName = $this->sanitiseClassName($this->modelClass);
        $gridFieldConfig = $form->Fields()->fieldByName($gridFieldName)->getConfig();

        $gridFieldConfig->removeComponentsByType('GridFieldPrintButton');
        $gridFieldConfig->removeComponentsByType('GridFieldExportButton');
        $gridFieldConfig
            ->getComponentByType('GridFieldAddNewButton')
            ->setButtonName(_t('Instagram.ButtonLabelAddAccount', 'Add account'));
        $gridFieldConfig
            ->getComponentByType('GridFieldDetailForm')
            ->setItemRequestClass('InstagramAdminRequestHandler');

        return $form;
    }

    /**
     * Gets a InstagramAccount ID from the Session.
     *
     * @param string $state
     * @return int|null
     */
    private function getInstagramAccountIDFromSession($state = null)
    {
        if (!$state) {
            return null;
        }

        $instagramAccounts = Session::get('InstagramAccounts');

        foreach ($instagramAccounts as $key => $value) {
            if ($value == $state) {
                return $key;
            }
        }
    }

    /**
     * Handles failed AOuth attempts.
     *
     * @param Form $form
     * @return Controller
     */
    private function handleOAuthError($form, $message = null)
    {
        $message = $message ? $message : _t(
            'Instagram.MessageOAuthErrorResponse',
            'Unable to authorise account. Please try again.'
        );

        $form->sessionMessage($message, 'bad');

        return Controller::curr()->redirect($this->Link());
    }

    /**
     * OAuth callback handler.
     *
     * @param SS_HTTPRequest $request
     */
    public function OAuth($request)
    {
        $code = $request->getVar('code');
        $state = $request->getVar('state');

        if (!$code || !$state) {
            return Controller::curr()->redirect($this->Link());
        }

        $client = InstagramAccount::getNewInstagramClient();
        $form = $this->getEditForm();

        try {
            $token = $client->getAccessToken($code);

            $instagramAccountID = $this->getInstagramAccountIDFromSession($state);

            // Find the matching InstagramAccount.
            if (!$instagramAccountID ||
                !$instagramAccount = InstagramAccount::get()->byId($instagramAccountID)
            ) {
                return $this->handleOAuthError($form);
            }

            try {
                $instagramAccount->updateAccessToken(Convert::raw2json($token), $state);
                $instagramAccount->write();

                $form->sessionMessage(
                    _t(
                        'Instagram.MessageOAuthSuccess',
                        'Successfully authorised your account.'
                    ),
                    'good'
                );

                return Controller::curr()->redirect($this->Link());
            } catch (Exception $e) {
                return $this->handleOAuthError($form, _t(
                    'Instagram.MessageOAuthErrorUserConflict',
                    'Unable to authorise account. Make sure you are logged out of Instagram and ' .
                    'your username is spelled correctly.'
                ));
            }
        } catch (InstagramIdentityProviderException $e) {
            return $this->handleOAuthError($form);
        }
    }
}

class InstagramAdminRequestHandler extends GridFieldDetailForm_ItemRequest
{
    private static $allowed_actions = [
        'edit',
        'view',
        'ItemEditForm'
    ];

    /**
     * @return Form
     */
    public function ItemEditForm()
    {
        $form = parent::ItemEditForm();

        $formActions = $form->Actions();

        if ($actions = $this->record->getCMSActions()) {
            foreach ($actions as $action) {
                $formActions->push($action);
            }
        }

        return $form;
    }
}


code/extension/InstagramPageExtension.php
<?php

use MetzWeb\Instagram\Instagram as Instagram;

class InstagramPageExtension extends DataExtension
{
    /**
     * @var array
     */
    private static $casting = [
        'instagramShortcodeHandler' => 'HTMLText',
    ];

    /**
     * Gets a list of media from the cache.
     *
     * @param InstagramAccount $instagramAccount
     * @param string $maxID
     * @return array|null
     */
    public static function getMediaFromCache($instagramAccount, $maxID)
    {
        $cache = SS_Cache::factory('InstagramPage');

        if (!($result = $cache->load("{$instagramAccount->getField('Title')}_{$maxID}"))) {
            return null;
        }

        return unserialize($result);
    }

    /**
     * Gets a list of media from Instagram.
     *
     * @param InstagramAccount $instagramAccount
     * @param string $maxID
     * @return array|null
     */
    public static function getMediaFromInstagram($instagramAccount, $maxID)
    {
        $response = $instagramAccount->getMedia($maxID);

        if (!$response) {
            return null;
        }

        $raw = $response->getRaw();

        if (!$raw || !array_key_exists('meta', $raw) || $raw['meta']['code'] !== 200) {
            user_error('Unable to fetch media from Instagram', E_USER_WARNING);
            return null;
        }

        return $raw['data'];
    }

    /**
     * Updates the cache with media fetched from Instagram.
     *
     * @param InstagramAccount $instagramAccount
     * @param string $maxID
     * @param array $media
     * @return null
     */
    public static function updateCachedMedia($instagramAccount, $maxID, $media)
    {
        $cache = SS_Cache::factory('InstagramPage');
        $cache->save(serialize($media), "{$instagramAccount->getField('Title')}_{$maxID}");
    }

    /**
     * Gets a list of media from the cache if available. Falls back to making an API request.
     *
     * @param InstagramAccount $instagramAccount
     * @param string $maxID
     * @return array|null
     */
    public static function getMedia($instagramAccount, $maxID) {
        // Return media from the cache if it's available.
        if ($media = self::getMediaFromCache($instagramAccount, $maxID)) {
            return $media;
        }

        // Fetch media from Instagram and cache it.
        if ($media = self::getMediaFromInstagram($instagramAccount, $maxID)) {
            self::updateCachedMedia($instagramAccount, $maxID, $media);
        }

        return $media;
    }

    /**
     * Handler for the `instagram` Shortcode.
     *
     * @return string|null
     */
    public static function instagramShortcodeHandler($args, $content = null, $parser = null, $tag)
    {
        if (!array_key_exists('username', $args)) {
            return null;
        }

        $instagramAccount = InstagramAccount::get()->filter('Title', $args['username'])->First();

        // Make sure the InstagramAccount is valid and has been authorised.
        if (!$instagramAccount || !$instagramAccount->getField('AccessToken')) {
            return null;
        }

        $controller = Controller::curr();
        $link = $controller->Link();

        // Valid IDs only have digits and an underscore.
        $maxID = preg_replace('/[^\d_]/', '', $controller->getRequest()->getVar('start'));

        $media = self::getMedia($instagramAccount, $maxID);

        if (!$media) {
            return null;
        }

        // If there are less items returned than the maximum allowed per page,
        // we're on the last page, so don't include a 'next' button.
        //
        // TODO:
        // Handle edge case where it's the last page and exactly "items_per_page" are returned.
        $loadMoreLink = count($media) < Config::inst()->get('InstagramAccount', 'items_per_page')
            ? null
            : $link . '?start=' . array_pop((array_slice($media, -1)))['id'];

        // Make tags iterable in the template.
        foreach ($media as &$item) {
            $item['tags'] = ArrayList::create(
                array_map(function ($tag) {
                    return ArrayData::create(['name' => $tag]);
                }, $item['tags'])
            );
        }

        $data = ArrayData::create([
            'loadMoreLink' => $loadMoreLink,
            'instagramLink' => "https://www.instagram.com/{$args['username']}"
        ]);

        $data->setField('media', ArrayList::create($media));

        return $data->renderWith('InstagramPage');
    }
}


code/model/InstagramAccount.php
<?php

use Larabros\Elogram\Client;

class InstagramAccount extends DataObject
{
    /**
     * @config
     * @var string
     */
    private static $client_id;

    /**
     * @config
     * @var string
     */
    private static $client_secret;

    /**
     * @config
     * @var string
     */
    private static $redirect_path;

    /**
     * @config
     */
    private static $items_per_page = 9;

    /**
     * @var array
     */
    private static $db = [
        'Title' => 'Varchar',
        'AccessToken' => 'Text',
    ];

    /**
     * @return FieldList
     */
    public function getCMSFields()
    {
        $fields = parent::getCMSFields();

        $fields->removeByName('AccessToken');

        if ($this->AccessToken) {
            $token = json_decode($this->AccessToken);

            $usernameField = LiteralField::create(
                'Title',
                '<div class="field">' .
                    '<label class="left">' .
                        _t('Instagram.FieldLabelTitle', 'Username') .
                    '</label>' .
                    '<div class="middleColumn" style="padding-top:8px;">' .
                        '<a ' .
                            "href='https://www.instagram.com/{$token->user->username}' " .
                            'title="View on Instagram" ' .
                            'target="_blank">' .
                            $token->user->username .
                        '</a>' .
                    '</div>' .
                '</div>'
            );
        } else {
            $usernameField = Textfield::create('Title', _t('Instagram.FieldLabelTitle', 'Username'));
            $usernameField->setDescription(
                _t(
                    'Instagram.FieldDescriptionTitle',
                    'The Instagram account you want to pull media from.'
                )
            );
        }

        $fields->addFieldToTab('Root.Main', $usernameField);

        $this->extend('updateCMSFields', $fields);

        return $fields;
    }

    /**
     * @return FieldList
     */
    public function getCMSActions()
    {
        $actions = parent::getCMSActions();

        if (!$this->ID || $this->AccessToken) {
            $this->extend('updateCMSActions', $actions);
            return $actions;
        }

        $client = self::getNewInstagramClient();
        $loginURL = $client->getLoginUrl();

        $this->setSessionOAuthState($this->getOAuthStateValueFromLoginURL($loginURL));

        $actions->push(
            LiteralField::create(
              'OAuthLink',
              '<a class="ss-ui-button" href="' . $loginURL . '">' .
                _t('Instagram.ButtonLabelAuthoriseAccount', 'Authorise account') .
              '</a>'
            )
        );

        $this->extend('updateCMSActions', $actions);

        return $actions;
    }

    /**
     * @return RequiredFields
     */
    public function getCMSValidator()
    {
        return new RequiredFields('Title');
    }

    /**
     * Ensures the person to authorising the account is logged into Instagram as the correct user
     * before setting the token.
     *
     * For example if the user is logged into Instagram as 'FooUser' and they attempt to set a token
     * for the InstagramAccount record 'InstagramAccount', the token will contain data relating to
     * the 'FooUser' account. So if this happen we display a message telling the user they need to
     * log out of Instagram before they can authorise another account.
     *
     * @param string $token
     */
    public function updateAccessToken($token, $state)
    {
        $newToken = json_decode($token);

        if ($state !== $this->getSessionOAuthState() ||
            $newToken->user->username !== $this->getField('Title')) {
            throw new Exception('Trying to set token on wrong InstagramAccount');
        }

        if (!$currentToken = $this->getField('AccessToken')) {
            $this->setField('AccessToken', $token);
            return;
        }

        $currentToken = json_decode($currentToken);

        if ($newToken->user->id !== $currentToken->user->id) {
            throw new Exception('Trying to set token on wrong InstagramAccount');
        }

        $this->setField('AccessToken', $token);
    }

    /**
     * Create a configured Instagram API interface.
     *
     * @param string $token
     * @return MetzWeb\Instagram\Instagram
     */
    public static function getNewInstagramClient($token = null)
    {
        $client_id = Config::inst()->get('InstagramAccount', 'client_id');
        $client_secret = Config::inst()->get('InstagramAccount', 'client_secret');
        $redirect_path = Config::inst()->get('InstagramAccount', 'redirect_path');

        if (!$client_id) {
            user_error(
                'Add a client_id to config (InstagramAdmin::client_id)',
                E_USER_ERROR
            );
        }

        if (!$client_secret) {
            user_error(
                'Add a client_secret to config (InstagramAdmin::client_secret)',
                E_USER_ERROR
            );
        }

        return new Client(
            $client_id,
            $client_secret,
            $token,
            Director::absoluteBaseURL() . $redirect_path
        );
    }

    /**
     * Gets the 'state' value from an OAuth login URL.
     *
     * @param string $loginURL
     * @return string|null
     */
    public function getOAuthStateValueFromLoginURL($loginURL = null)
    {
        if (!$loginURL) {
            return null;
        }

        $parts = parse_url($loginURL);
        parse_str($parts['query'], $query);

        return array_key_exists('state', $query)
            ? $query['state']
            : null;
    }

    /**
     * Gets the InstsgramAccount's OAuth state from Session.
     *
     * @return string|null
     */
    public function getSessionOAuthState()
    {
        $instagramAccounts = Session::get('InstagramAccounts');

        if (!$this->ID || !$instagramAccounts || !array_key_exists($this->ID, $instagramAccounts)) {
            return null;
        }

        return $instagramAccounts[$this->ID];
    }

    /**
     * Gets the authorised user's Instagram ID from the AccessToken.
     *
     * @return string|null
     */
    public function getInstagramID()
    {
        if (!$token = $this->getField('AccessToken')) {
            return null;
        }

        return json_decode($token)->user->id;
    }

    /**
     * Checks if the passed ID is a valid pattern.
     *
     * @param string $mediaID
     * @return boolean
     */
    public function isValidMediaID($mediaID = null)
    {
        if (!$mediaID || !$instagramID = $this->getInstagramID()) {
            return false;
        }

        $pattern = '/^(\d{18}|\d{19})_' . $instagramID . '$/';

        return preg_match($pattern, $mediaID) === 1;
    }

    /**
     * Sets a Session variable which is used to keep track of
     * the InstagramAccount through the OAuth flow.
     *
     * @param string $state
     */
    private function setSessionOAuthState($state = null)
    {
        if (!$this->ID || !$state) {
            return null;
        }

        $instagramAccounts = Session::get('InstagramAccounts');
        $instagramAccounts = $instagramAccounts ? $instagramAccounts : [];

        $instagramAccounts[$this->ID] = $state;

        Session::set('InstagramAccounts', $instagramAccounts);
    }

    /**
     * Gets a list of media from the user's Instagram.
     *
     * @param string $maxID Return media earlier than this ID
     * @return Larabros\Elogram\Http\Response|null
     */
    public function getMedia($maxID = null)
    {
        if (
            !is_string($this->getField('AccessToken')) ||
            ($maxID && !$this->isValidMediaID($maxID))
        ) {
            return null;
        }

        $client = $this->getNewInstagramClient($this->getField('AccessToken'));
        $client->secureRequests();

        return $client->users()->getMedia('self', $this->config()->items_per_page, null, $maxID);
    }
}
